providers/implementations/signature/dsa_sig.c

```diff
@@ -562,6 +578,190 @@ static const OSSL_PARAM *dsa_settable_ctx_md_params(void *vpdsactx)
     return EVP_MD_settable_ctx_params(pdsactx->md);
 }
 
+#ifdef FIPS_MODULE
+
+static CRYPTO_RWLOCK *selftest_lock;
+
+static void selftest_lock_init(void)
+{
+    selftest_lock = CRYPTO_THREAD_lock_new();
+}
+
+static int dsa_digest_signverify_selftest(OSSL_LIB_CTX *libctx)
+{
+    static int selftest_status = -1;
+    static CRYPTO_ONCE selftest_once = CRYPTO_ONCE_STATIC_INIT;
+    static CRYPTO_THREAD_ID selftest_thread = {0};
+    int success = 0;
+    size_t i;
+    const char *msg = "Hello World!";
+    unsigned char sig[256];
+    size_t siglen = 0;
+    /* dsa_newctx() and dsa_newdata() both need a provctx, but only use libctx
+     * from it. We don't have a provctx here, but libctx is easy to obtain.
+     * Create a test PROV_CTX that just has the libctx field set. */
+    PROV_CTX test_provider = {0};
+    PROV_DSA_CTX *ctx = NULL;
+    DSA *provkey = NULL;
+
+    const OSSL_DISPATCH *fns = NULL;
+    OSSL_FUNC_keymgmt_new_fn *dsa_newdata = NULL;
+    OSSL_FUNC_keymgmt_free_fn *dsa_freedata = NULL;
+    OSSL_FUNC_keymgmt_import_fn *dsa_import = NULL;
+
+    OSSL_PARAM_BLD *bld = NULL;
+    const ST_KAT_PARAM *p = NULL;
+    BN_CTX *bnctx = NULL;
+    OSSL_PARAM *key = NULL;
+
+    if (selftest_status != -1)
+        return selftest_status;
+
+    /* If this is the thread that is currently running the selftest, don't
+     * attempt to lock again, just succeed. */
+    if (CRYPTO_THREAD_compare_id(selftest_thread,
+                                 CRYPTO_THREAD_get_current_id()))
+        return 1;
+
+    if (!CRYPTO_THREAD_run_once(&selftest_once, selftest_lock_init)
+        || selftest_lock == NULL)
+        return 0;
+
+    if (!CRYPTO_THREAD_write_lock(selftest_lock))
+        return 0;
+
+    selftest_thread = CRYPTO_THREAD_get_current_id();
+
+    test_provider.libctx = libctx;
+
+    for (fns = ossl_dsa_keymgmt_functions; fns->function_id != 0; fns++) {
```

Comment:
Since this is inside the FIPS provider, why not just call the relevant functions directly?  External implementations aren't possible and we know what the contents of that table are.